src/main/scala/tech/sourced/gemini/FileQuery.scala

```diff
@@ -0,0 +1,153 @@
+package tech.sourced.gemini
+
+import java.io.File
+import java.nio.file.Files
+
+import com.datastax.driver.core.{Session, SimpleStatement}
+import gopkg.in.bblfsh.sdk.v1.uast.generated.Node
+import org.bblfsh.client.BblfshClient
+import org.slf4j.{Logger => Slf4jLogger}
+import tech.sourced.featurext.FEClient
+import tech.sourced.featurext.generated.service.Feature
+import tech.sourced.featurext.generated.service.FeatureExtractorGrpc.FeatureExtractor
+
+import scala.collection.JavaConverters._
+import scala.util.control.NonFatal
+
+/**
+  * QueryResult contains iterators for full duplicates and similar files
+  *
+  * @param duplicates
+  * @param similar
+  */
+case class QueryResult(duplicates: Iterable[RepoFile], similar: Iterable[RepoFile])
+
+/**
+  * FileQuery queries similar and duplicated files
+  *
+  * @param conn db connection
+  * @param bblfshClient
+  * @param feClient
+  * @param docFreqPath will be replaced by database in future
+  * @param log
+  * @param keyspace
+  * @param tables
+  */
+class FileQuery(conn: Session,
+            bblfshClient: BblfshClient,
+            feClient: FeatureExtractor,
+            docFreqPath: String = "",
+            log: Slf4jLogger,
+            keyspace: String,
+            tables: Tables) {
+
+  /**
+    * find similar and duplicates for a file
+    *
+    * @param file
+    * @return
+    */
+  def find(file: File): QueryResult = {
+    val duplicates = findDuplicatesOfFile(file)
+    val duplicatedShas = duplicates.map(_.sha).toSeq
+
+    val similarShas = findSimilarForFile(file, conn, bblfshClient, feClient, docFreqPath)
+      .map(_.split("@")(1)) // value for sha1 in Apollo hashtables is 'path@sha1', but we need only hashes
+      .filterNot(sha => duplicatedShas.contains(sha))
+
+    log.info(s"${similarShas.length} SHA1's found to be similar")
+
+    val similar = similarShas.flatMap(sha1 => Database.repoFilesByHash(sha1, conn, keyspace, tables))
+    QueryResult(duplicates, similar)
+  }
+
+  protected def findDuplicatesOfFile(file: File): Iterable[RepoFile] = {
+    Database.repoFilesByHash(Gemini.computeSha1(file), conn, keyspace, tables)
+  }
```

Comment:
It seems to me that this guy can be safely removed in favor of `Database.findFilesByHash()` or something like that.